Register Action does not currently display error messages to the user

mdbitz opened the pull request:
Register Action does not currently display error messages to the user. This is causing confusion as users repeatedly try to 'Set up Jetpack' and are always shown the same screen without any info on the error experienced.

We are correcting this by changing  to ->error to ensure the admin_notices are rendered.

Note that pretty error messages are needed for all error codes.

Fixes #

#### Changes proposed in this Pull Request:
* Update Register logic to display error messages to the user.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Updates existing Connection Flow

#### Testing instructions:
Current Behavior is shown with
* Disable support for XML-RPC or mark the site private
* Activate Jetpack on your site
* Try "Set up Jetpack"  (/wp-admin/admin.php?page=jetpack&action=register&_wpnonce=XXXXXX&from=full-screen-prompt)
* User will be shown the same "Set up Jetpack" screen 

Apply Patch
* Try "Set up Jetpack" . 
* User will see an admin notice with the error experienced

#### Proposed changelog entry for your changes:
* No Changelog needed

mdbitz committed:
Register Action does not currently display error messages to the user. This is causing confusion as users repeately try to 'Set up Jetpack' and are always shown the same screen without any info on the error experienced.

We are correcting this by changing  to ->error to ensure the admin_notices are rendered.

Note that pretty error messages are needed for all error codes.

mdbitz committed:
Port Error messages from jsx files.

mdbitz committed:
Formatting

dereksmart commented on class.jetpack.php:
```diff
@@ -4260,18 +4260,143 @@ function admin_page_load() {
+	 *
+	 * @return string
+	 */
+	private function pretty_error_codes( $error ) {
```
Does moving these to PHP mean that we can also remove them from the .jsx file?

mdbitz replied:
I'm unsure on this part, haven't fully explored the react app to say whether there are any plan changes or other actions that would trigger errors to show in react.

kraftbj replied:
Without comment on how to do it (an endpoint, passing this through as vars and `wp_localize_script`, etc), I would like us to only have one source of truth for error->messaging. We've been bitten before by having two lists for the same thing in separate places get out of sync.

mdbitz committed:
Corrected sprintf format

mdbitz committed:
Indenting changes only, need to get my IDE settings updated.

mdbitz committed:
Enabling Private site plugins triggers  error code. Updating the case to align with private site message

mdbitz committed:
Addition of support link to error message

mdbitz commented:
Notes:

This should be thoroughly tested, also we may want to modify the messages with additional self help links if they exist. I added contact support so far but if we have other docs to point users to 👍

kraftbj commented:
Moving this to 8.3 as part of properly fixing this. Per conversation with Crew, we are going to revert the no-app change in 7.9 that led to React (and thus the error messaging there) from not working for 8.1.1.

kbrown9 committed:
Connection: move error messages to a new class

- Create a new class in the connection package for handling connection error
  messages.
- Add a method in the Connection\Manager class which consumers can use to
  convert an error code to an error message.
- Use the above method in Jetpack.

kbrown9 committed:
Connection_Error_Handler: add comment for translators

zinigor commented on packages/connection/src/class-connection-error-handler.php:
```diff
@@ -0,0 +1,130 @@
+ * The Jetpack Connection Error Handler class is used to handle the errors
+ * generated during the connection process.
+ */
+class Connection_Error_Handler {
```
One small nitpick: can we call this class just `Error_Handler`? It's already a Connection package :)

kbrown9 replied:
Yeah - good point! I'll change it.

kbrown9 committed:
Connection: rename Connection_Error_Handler class to Error_Handler

zinigor commented on packages/connection/src/class-manager.php:
```diff
@@ -2167,4 +2167,15 @@ public function sign_role( $role, $user_id = null ) {
+	 * @return string The error message.
+	 */
+	public function get_error_message_from_code( $error_code ) {
+		return ( new Connection_Error_Handler() )->get_error_message_from_code( $error_code );
```
This will need renaming too now, won't it? I tried to find where it was `use`d, but it's not, so it's going to be accessed as `\Automattic\Jetpack\Connection\Connection_Error_Handler`, which doesn't exist anymore.

mdbitz commented:
Closing PR. This was created to resolve the issue of silent errors but was handled instead by https://github.com/Automattic/jetpack/pull/14373
